libs/cache/cache.go

```diff
@@ -0,0 +1,80 @@
+package cache
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"sort"
+)
+
+// Cache provides an abstract interface for caching content to local disk.
+// Implementations should handle storing and retrieving cached components
+// using fingerprints for cache invalidation.
+type Cache[T any] interface {
+	// GetOrCompute retrieves cached content for the given fingerprint, or computes it using the provided function.
+	// If the content is found in cache, it is returned directly.
+	// If not found, the compute function is called, its result is cached, and then returned.
+	// The fingerprint can be any struct that will be serialized deterministically for cache key generation.
+	// Returns an error if the cache operation or compute function fails.
+	GetOrCompute(ctx context.Context, fingerprint any, compute func(ctx context.Context) (T, error)) (T, error)
+}
+
+// fingerprintToHash converts any struct to a deterministic string representation for use as a cache key.
+func fingerprintToHash(fingerprint any) (string, error) {
+	// Serialize to JSON with sorted keys for deterministic output
+	data, err := json.Marshal(fingerprint)
```

Comment:
We seem to call `normalizeForFingerprint` for normalisatin and key ordering, why marshal/unmarshal?